src/shogun/base/class_list.cpp.py

```diff
@@ -67,80 +72,86 @@ def extract_class_name(lines, line_nr, line, blacklist):
 
     c = c.strip(':').strip()
 
-    if not c.startswith('C'):
+    if not is_shogun_class(c):
         return
     if c.endswith(';'):
         return
     if '>' in c:
         return
-    if not (len(c) > 2 and c[1].isupper()):
-        return
     if check_is_in_blacklist(c[1:], lines, line_nr, blacklist):
         return
 
     return c[1:]
 
 
-def get_includes(classes, basedir="."):
-    class_headers = []
-    for c, t in classes:
-        class_headers.append(c+".h")
+def extract_base_class_name(lines, line_nr, line):
+    if not public_str in line:
+        line += lines[line_nr+1]
+    s = line[line.index(public_str)+len(public_str):]
+    if '<' in s:
+        b = s[:s.index('<')]
+    else:
+        b = s.split()[0]
+    b = b.strip(' {')
+    if is_shogun_class(b):
+        return b[1:]
 
-    import os
-    result = []
-    for root, dirs, files in os.walk(basedir):
-        for f in files:
-            if f in class_headers:
-                result.append(os.path.join(os.path.relpath(root, basedir), f))
 
-    includes = []
-    result.sort()
-    for o in result:
-        includes.append('#include <shogun/%s>' % o.strip().lstrip('./'))
-    return includes
+def get_includes(classes):
+    includes = set()
+    for attrs in classes.values():
+        if not attrs['abstract']:
+            includes.add('#include <shogun/%s>' % attrs['header'])
```

Comment:
@lisitsyn @vigsterkr I guess we could also use ctags here, just as for the meta examples?